ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/anomalylocalization/AnomalyLocalizerImpl.java

```diff
@@ -0,0 +1,381 @@
+
+import org.opensearch.action.ActionListener;
+import org.opensearch.action.NotifyOnceListener;
+import org.opensearch.action.search.MultiSearchRequest;
+import org.opensearch.action.search.MultiSearchResponse;
+import org.opensearch.action.search.SearchRequest;
+import org.opensearch.action.search.SearchResponse;
+import org.opensearch.client.Client;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.index.query.BoolQueryBuilder;
+import org.opensearch.index.query.QueryBuilders;
+import org.opensearch.index.query.RangeQueryBuilder;
+import org.opensearch.index.query.TermQueryBuilder;
+import org.opensearch.search.aggregations.AggregationBuilder;
+import org.opensearch.search.aggregations.AggregationBuilders;
+import org.opensearch.search.aggregations.bucket.composite.CompositeAggregation;
+import org.opensearch.search.aggregations.bucket.composite.CompositeAggregationBuilder;
+import org.opensearch.search.aggregations.bucket.composite.TermsValuesSourceBuilder;
+import org.opensearch.search.aggregations.bucket.filter.Filters;
+import org.opensearch.search.aggregations.bucket.filter.FiltersAggregationBuilder;
+import org.opensearch.search.aggregations.bucket.filter.FiltersAggregator.KeyedFilter;
+import org.opensearch.search.aggregations.metrics.NumericMetricsAggregation.SingleValue;
+import org.opensearch.search.builder.SearchSourceBuilder;
+
+import lombok.Data;
+import lombok.SneakyThrows;
+import lombok.extern.log4j.Log4j2;
+
+import static org.opensearch.action.ActionListener.wrap;
+import static org.opensearch.search.aggregations.MultiBucketConsumerService.MAX_BUCKET_SETTING;
+
+/**
+ * Implementation of AnomalyLocalizer.
+ */
+@Log4j2
+public class AnomalyLocalizerImpl implements AnomalyLocalizer {
+
+    protected static final double MIN_DIFF_PCT = 0.01;
+    protected static final int MAX_TIME_BUCKETS = 8;
+
+    private final Client client;
+    private final Settings settings;
+
+    /**
+     * Constructor.
+     *
+     * @param client   Index data.
+     * @param settings Settings information.
+     */
+    public AnomalyLocalizerImpl(
+            Client client,
+            Settings settings) {
+        this.client = client;
+        this.settings = settings;
+    }
+
+    /**
+     * Implementation of localization.
+     * <p>
+     * This localization finds the largest contributors to changes in aggregated data over time.
+     * The entire time range is first partitioned into equally sized buckets.
+     * Each bucket of data is aggregated as a whole and also sliced into entity key-value pairs. With first bucket being the base,
+     * the change from a new bucket later is the difference of the new aggregate value and the base aggregate value.
+     * The contribution to the change from an entity is the difference between its new value and base value.
+     */
+    @Override
+    @SneakyThrows
+    public void getLocalizationResults(Input input, ActionListener<Output> listener) {
```

Comment:
nit: add `final` keyword whenever it's applicable.  
e.g.,`(final Input input, final ActionListener<Output> listener)`

Reply:
i would suggest no for a few reasons.
1. it doesn't improve code readability, just imagine lots of final keywords in almost every line of code. good implementation is correct and the way to verify is via testing, (think of other general programming languages such as javascript, python). specific language constructs help, more general better approach is static code analysis and code style checks and other automated tools (such as code guru).
2. There are abundant examples or convention in opensearch. 
https://github.com/opensearch-project/OpenSearch/blob/996d33adb22ac6a523962166f3677f8dc6ac9c1f/server/src/main/java/org/opensearch/action/search/SearchTransportService.java#L109
https://github.com/opensearch-project/OpenSearch/blob/996d33adb22ac6a523962166f3677f8dc6ac9c1f/server/src/main/java/org/opensearch/search/internal/ReaderContext.java#L60